### ISSUE TITLE
Building with openssl 3.0.5, issue with RSA_set0_key() in p11_rsa.c

### ISSUE THREAD

**Issue by jjeffroberts** [OpenSC/libp11#481]
We are working through an upgrade and have picked up libp11 0.4.11 and openssl 3.0.5

When we look up a private RSA key, I've traced down to src/p11_rsa.c:pkcs11_get_rsa(PKCS11_KEY *key)

I can see where an (RSA *) is created with RSA_new() and then this call:

RSA_set0_key(rsa, rsa_n, rsa_e, NULL);

Prior to upgrades, this ultimately yielded a key that I could send to EVP_PKEY_decrypt() but that now fails.

When I call ERR_print_errors() I see 'missing private key'.

If I dump the key with PEM_write_RSAPrivateKey() and PEM_write_RSAPublicKey(), both return 0 but PEM_write_RSAPrivateKey() prints nothing.

I made some changes to see what would happen if I set CKA_PRIVATE_EXPONENT when I store the key and then look it up in pkcs11_get_rsa() and pass it to RSA_set0_key(). This does allow me to call EVP_PKEY_decrypt() successfully, but PEM_write_RSAPrivateKey() still returns 0 but prints nothing, oddly.

Have you tested with openssl 3?

**Comment by dengert** [comment]
See https://github.com/OpenSC/libp11/issues/432 and comments https://github.com/OpenSC/libp11/issues/432#issuecomment-1303545558 and 
 https://github.com/OpenSC/libp11/issues/432#issuecomment-1301998777

IMHO, libp11, without provider support will die when OpenSSL drops support  for engines.

https://github.com/latchset/pkcs11-provider/ will replace it.

**Comment by mouse07410** [comment]
> IMHO, libp11, without provider support will die when OpenSSL drops support for engines.
> https://github.com/latchset/pkcs11-provider/ will replace it.

Currently, `pkcs11-provider` is unusable on MacOS. It even fails to compile, let alone do anything else. I'm working on getting to compile reliably.

**Comment by mtrojnar** [comment]
There seems to be a regression in the latest releases of the OpenSSL stable branch.  It used to work with older OpenSSL 3.0 releases.  Now, not only libp11 is affected, but also pkcs11-tool fails with the "OpenSSL error during RSA private key parsing" error.
@olszomal Could you investigate it?

**Comment by jjeffroberts** [comment]
FYI, I have opened this with openssl:

https://github.com/openssl/openssl/issues/20150

Thanks,
-Jeff

**Comment by dengert** [comment]
Have you tried building from https://github.com/openssl/openssl ?

**Comment by jjeffroberts** [comment]
I created the example that I attached to the openssl issue by cloning from github, checking out their openssl-3.0 branch, doing a build and test to make sure I was in a good state, then adding that example as a new test in test/rsa_test.c

So far their responses are that this could never have worked. My plan today is to check out their 1.1.1 branch and see if I can show that it worked.

**Comment by jjeffroberts** [comment]
So far I have not been able to get a good private key in the example when I run it against openssl 1.1.1 on my laptop. I've tried the code on their 1.1.1 stable branch and at the 1.1.1q tag.

The openssl version in our actual (working) build is 1.1.1q plus some patches, most of which were CVE fixes from upstream openssl.

I'm going to start removing patches to see if I can get it to fail.

**Comment by dengert** [comment]
Do you have a test command that shows the failure?

Are you using smartcards too?

**Comment by jjeffroberts** [comment]
The short answer is no, and I'm not sure how I would get from what we have to something simpler.

A lot of this pre-dates me so pardon if I'm inaccurate in my description.

We use libp11 as part of a mock that simulates a card for unit tests. We read in certs and keys and load them into libp11 and then eventually our unit tests make calls like PKCS11_enumerate_keys(). I can see where we're setting up for a decrypt and we want the private key and ultimately we call p11_rsa.c:pkcs11_get_rsa(), which is where the call to RSA_set0_key(rsa, rsa_n, rsa_e, NULL); is made.

In my local build I can pretty easily change our front end to save the CKA_PRIVATE_EXPONENT as we're loading the key, so that it's there to get via key_getattr_bn() once I get to pkcs11_get_rsa(). I could get the tests working with that if nothing comes of this.

I could probably chalk this all up to cross-compiled unit test code at build time or optimization or something, but then I would think that call would be failing for anyone else using libp11 to decrypt with RSA keys, since our "side-effect" is that it actually works.

**Comment by dengert** [comment]
Just looked at the https://github.com/openssl/openssl/issues/20150 Their answers are correct. Without "d" its a public key. 

pkcs11_get_rsa is called to get only e and n.  PKCS11 has both public keys and private keys objects. OpenSSL RSA is a combination of both. So  problem is most likely https://github.com/OpenSC/libp11/blob/master/src/p11_rsa.c#L287-L292 
and https://github.com/OpenSC/libp11/blob/master/src/p11_rsa.c#L251-L254

https://www.openssl.org/docs/man3.1/man3/RSA_set_ex_data.html
"All functions with a TYPE of DH, DSA, RSA and EC_KEY are deprecated. Applications should instead use EVP_PKEY_set_ex_data(), EVP_PKEY_get_ex_data() and EVP_PKEY_get_ex_new_index()."  

In other words the hooks libp11 uses to get control to do crypto operations on a smartcard (where "d") is never exposed are based on adding hooks to OpenSL objects. Only the EVP_PKEY objects support these hooks. 

Unfortunately  many of the libp11 calls do not check for errors. And I don't believe many programs uses  _ex_data so they may not even work these days. 

 You could try using  LibreSSL to see if it works.

**Comment by dengert** [comment]
If you are only using libp11 as part of a mock that simulates a card for unit tests. You could look at how OpenSC tests its code using virtual card reader with a java program that runs card applets for a number of cards. 
 Under: https://github.com/OpenSC/OpenSC/tree/master/.github    workflow/linux.yml  runs  test-openpgp.sh

**Comment by jjeffroberts** [comment]
Your analysis makes sense to me for openssl 3.x. I'm not sure that I understand why it doesn't always seem to work for openssl 1.x. Are there any compile-time flags that might enable/disable the _ex functions in earlier openssl?

Thanks for looking into this. I have some options now.

Do I close this or do you?

**Comment by dengert** [comment]
No flags to enable  _ex functions. 

You can add CFLAGS= -6  and LD_FLAGS=-g  to your program,  libp11 and  OpenSSL so if you run `gdb --args your command and its args`  you can set break points,  show the call stack and see source lines, and look at variables. 

You can close it.

**Comment by jjeffroberts** [comment]
I keep almost dropping this, but now that you've mentioned those methods... I think in our old build we go through pkcs11_rsa_priv_dec_method() and in our new build we do not.

This stuff is a bit too convoluted to debug real-time, I'm stuck with logging but maybe I have a better idea of what I'm looking for now.

**Comment by jjeffroberts** [comment]
I think I'm starting to get my head around it now.

In pkcs11_get_evp_key_rsa(), if key->isPrivate there is a call to RSA_set_method() that hooks in libp11's versions of various methods, including decrypt. And of course I never call any of that when I'm just calling openssl myself in my example.

In my old build (with openssl 1.1.1q) I can see that when we call EVP_PKEY_decrypt(), libp11's pkcs11_rsa_priv_dec_method() is called, and it goes on to call the _ex_ methods you mentioned.

In my new build (with openssl 3.0.5) I don't even think the RSA_set_method() is working - I see no call to pkcs11_rsa_priv_dec_method() when I call EVP_PKEY_decrypt().

I think I'm going to stop here.

**Comment by jjeffroberts** [comment]
Narrator: He did not stop there.

I managed to create a small testcase for this. I just borrowed some of your code to create a custom decrypt method, hook it in with RSA_set_method() and then call EVP_PKEY_decrypt() to see if it calls the custom method.

It calls the method for openssl 1.1.1q but not for the code on their openssl-3.0 branch, at their openssl-3.0.5 tag, or on their master branch.

I realize some of the methods are deprecated, but they should still work for now - they define deprecated as:

> You may still use them in your applications, but you may start to see deprecation warnings during compilation (dependent on compiler support for this). Deprecated APIs may be removed from future versions of OpenSSL so you are strongly encouraged to update your code to use the high level APIs instead.

I'm going to clean this up a little and log a new issue with openssl.

**Comment by jjeffroberts** [comment]
FYI, it's this: https://github.com/openssl/openssl/issues/20176

It has been marked "triaged: bug".

**Comment by mouse07410** [comment]
I've a naive question. Is the problem just one specific function call that the current combination of `libp11` and OpenSSL-3.0.x fails? Asking because `libp11` seems to work fine for me with OpenSSL-3.0.8 and OpenSSL-3.2.x-dev, both with CLI and my own apps that invoke libp11 engine:

```
Using YubiKey NEO device

Data to sign would be 83 bytes long (not equal to 32 bytes aka size of SHA256)

Using OpenSSL-3.x

Enter PKCS#11 token PIN for XXXXXXXXXXXX: xxxxxxxxx

Sign/verify using RSA-PSS
Signed (data len=83, signature 256 bytes): rv=1
23 f1 2c 85 b9 d4 f3 0a 68 74 4e 55 75 11 88 ab 
c4 0a d2 4c 96 69 59 60 78 b8 3b ac 7b 4f e6 8d 
23 90 0a 05 08 80 de e0 bf ec e5 56 6c 2f 40 6e 
de b7 a2 69 43 8e ec 8e cd e1 ee cc de 11 8b e5 
a1 7c ff bb 4d b3 4d 89 21 dc 4b 6a b1 29 c1 9f 
e4 1e 74 ba 50 9e 57 f8 9d de d5 88 ec 61 e6 3a 
8e 2a 21 6e 8d 87 f5 a4 c7 df 6f b9 df 62 40 90 
39 d8 db 73 95 07 a2 db 4b 5d 14 94 c2 85 86 ff 
39 a8 11 d5 c8 f1 88 63 77 18 ac 46 73 d1 fd f2 
ea 6d 02 2a 88 28 6c 03 60 8d 24 88 df 55 0c 56 
ca f5 9b a9 e8 39 43 16 71 0e 5f 8b a9 1a 89 99 
d4 51 54 d8 c9 32 4f 00 21 d9 ac 49 86 ae 1d 06 
40 d4 bd 9b 62 45 f2 bb 96 6a 23 c2 64 0b 36 01 
ae 1c 2d 7a 41 ea f3 de da 1b 89 2c 4f 8d 9d 3b 
10 ce 8e 14 c9 d3 8b 40 a0 98 ba a8 2f 12 7b fb 
64 35 2b 07 9a d5 28 78 ad e0 be f6 74 69 12 03 

RSA-PSS signature verified successfully: rv=1
```
and
```
$ pkcs11-rsa-pss-sign-demo2
This is not a CAC
Generating ephemeral file /tmp/derive.72593.text to test RSA-PSS signature...

openssl rand -engine rdrand -hex -out /tmp/derive.72593.text 5120
Engine "rdrand" set.

Signing file /tmp/derive.72593.text...
openssl dgst -engine pkcs11 -keyform engine -sign "pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;id=%02;object=SIGN%20key;type=private" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -out /tmp/derive.72593.text.sig /tmp/derive.72593.text
Engine "pkcs11" set.
Enter PKCS#11 token PIN for XXXXXXXXXXXX:
Signature for /tmp/derive.72593.text is stored in /tmp/derive.72593.text.sig

Verifying signature:
openssl dgst -engine pkcs11 -keyform engine -verify "pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;id=%02;object=SIGN%20pubkey;type=public" -sha384 -sigopt rsa_padding_mode:pss -sigopt rsa_pss_saltlen:-1 -signature /tmp/derive.72593.text.sig  /tmp/derive.72593.text
Engine "pkcs11" set.
Verified OK

Verifying signature using pkcs11-tool:
pkcs11-tool --module /usr/local/lib/opensc-pkcs11.so --verify -m SHA384-RSA-PKCS-PSS -y cert -d 02 --signature-file /tmp/derive.72593.text.sig -i /tmp/derive.72593.text
Using slot 1 with a present token (0x4)
Using signature algorithm SHA384-RSA-PKCS-PSS
PSS parameters: hashAlg=SHA384, mgf=MGF1-SHA384, salt_len=48 B
Signature is valid

$ pkcs11-rsa-encr-demo2
This is not a CAC
Using KEY MAN key...
Generate a random data file (1000 bytes), Base64-encoded...
openssl-3 rand -engine rdrand -base64 -out /tmp/derive.72980.text 1000
Engine "rdrand" set.

Engine "rdrand" set.
Engine "rdrand" set.
Generated
  key: 8f099a88f2f99b29d81a3d89eb567f71289cb6050789baef1951781a5ccca5b6
  iv:  7fd1d341e75ed366585714c0a2f7216e

Encrypt file /tmp/derive.72980.text with this key and IV...
openssl-3 enc -aes-256-cfb -a -e -in /tmp/derive.72980.text -out /tmp/derive.72980.text.enc -K 8f099a88f2f99b29d81a3d89eb567f71289cb6050789baef1951781a5ccca5b6 -iv 7fd1d341e75ed366585714c0a2f7216e

Encrypting this random symmetric key to token RSA KEY MAN key...
echo 8f099a88f2f99b29d81a3d89eb567f71289cb6050789baef1951781a5ccca5b6 | xxd -r -p -c 200 | openssl-3 pkeyutl -engine pkcs11 -keyform engine -encrypt -pubin -inkey "pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;id=%03;object=KEY%20MAN%20pubkey;type=public" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -out /tmp/derive.72980.key.enc
Engine "pkcs11" set.

Decrypting the symmetric key on the token...
openssl-3 pkeyutl -engine pkcs11 -keyform engine -decrypt -inkey "pkcs11:model=PKCS%2315%20emulated;manufacturer=piv_II;id=%03;object=KEY%20MAN%20key;type=private" -pkeyopt rsa_padding_mode:oaep -pkeyopt rsa_oaep_md:sha384 -pkeyopt rsa_mgf1_md:sha384 -in /tmp/derive.72980.key.enc | xxd -p -c 200
Engine "pkcs11" set.
Enter PKCS#11 token PIN for XXXXXXXXXXXXXX:

Decrypt file /tmp/derive.72980.text.enc with this key and IV...
openssl-3 enc -aes-256-cfb -a -d -in /tmp/derive.72980.text.enc -out /tmp/derive.72980.text.dec -K 8f099a88f2f99b29d81a3d89eb567f71289cb6050789baef1951781a5ccca5b6 -iv 7fd1d341e75ed366585714c0a2f7216e

KEY1="8f099a88f2f99b29d81a3d89eb567f71289cb6050789baef1951781a5ccca5b6"
KEY2="8f099a88f2f99b29d81a3d89eb567f71289cb6050789baef1951781a5ccca5b6"

Original and decrypted keys  match

Original and decrypted files match


Cleaning up...


$ 
```

**Comment by jjeffroberts** [comment]
I got a reponse from the openssl team and this re-ordering of calls gets things working again:

diff --git a/src/p11_rsa.c b/src/p11_rsa.c
index 6356f7c..760ef5a 100644
--- a/src/p11_rsa.c
+++ b/src/p11_rsa.c
@@ -309,7 +309,6 @@ static EVP_PKEY *pkcs11_get_evp_key_rsa(PKCS11_KEY *key)
 		RSA_free(rsa);
 		return NULL;
 	}
-	EVP_PKEY_set1_RSA(pk, rsa); /* Also increments the rsa ref count */
 
 	if (key->isPrivate) {
 		RSA_set_method(rsa, PKCS11_get_rsa_method());
@@ -319,6 +318,10 @@ static EVP_PKEY *pkcs11_get_evp_key_rsa(PKCS11_KEY *key)
 		rsa->flags |= RSA_FLAG_EXT_PKEY;
 #endif
 	}
+
+	// This must be set AFTER we install the custom methods.
+	EVP_PKEY_set1_RSA(pk, rsa); /* Also increments the rsa ref count */
+
 	/* TODO: Retrieve the RSA private key object attributes instead,
 	 * unless the key has the "sensitive" attribute set */
